flax/linen/module.py

```diff
@@ -0,0 +1,415 @@
+class Module:
+  """Base Module Class"""
+  # class defaults -- Note: these __must not__ be made annotations!
+  _in_compact_method = False
+  _in_call = False
+  _in_setup = False
+  _last_varname = None
+  scope = None
+
+  @classmethod
+  def __init_subclass__(cls):
+    """Automatically initialize all subclasses as custom dataclasses."""
+    # All Flax Modules are dataclasses.  We force this convention since
+    # it encourages the stateless behavior needed to clone module instances for
+    # functional transformation.  Instead of using a python metaclass, we
+    # automatically transform Modules into dataclasses at subclass creation
+    # time, and we set the first dataclass argument to `parent`, and the last
+    # optional argument to `name`.
+    cls._add_parent_and_name_attrs()
+    dataclasses.dataclass(cls)
+    cls.setup = wrap_setup(cls.setup)
+
+  @classmethod
+  def _add_parent_and_name_attrs(cls):
+    """Add final optional dataclass attributes: `parent` and `name`."""
+    annotations = cls.__dict__.get('__annotations__', {})
+    if 'parent' in annotations or 'name' in annotations:
+      raise ValueError(
+          f'properties `parent` and `name` are reserved: {annotations}')
+    # Add `parent` and `name` default fields at end.
+    new_annotations = {}
+    new_annotations.update(annotations)
+    if 'parent' in getattr(cls, '__dataclass_fields__', {}):
+      cls.__dataclass_fields__.pop('parent')
+    new_annotations['parent'] = Union[Type["Module"], Type["Scope"],
+                                      Type["Sentinel"], None]
+    cls.parent = dataclasses.field(repr=False, default=_unspecified_parent)
+    if 'name' in getattr(cls, '__dataclass_fields__', {}):
+      cls.__dataclass_fields__.pop('name')
+    new_annotations['name'] = str
+    cls.__annotations__ = new_annotations
+    cls.name = None  # default value of name is None.
+
+  def __setattr__(self, name: str, val: Any):
+    """We overload setattr solely to support pythonic naming via
+    assignment of submodules in the special setup() function:
+      self.submodule_name = MyModule(...)
+    we also support lists and other general pytrees, e.g.:
+      self.submodules = [MyModule0(..), MyModule1(..), ...]
+    """
+    # val is a Module or pytree whose leaves are all Modules.
+    if is_module_tree(val):
+      # We don't mess with the parent module.
+      if name == 'parent':
+        pass
+      # Modules have been passed in as dataclass args.
+      # if they're orphaned, i.e. Module(None, ...) then attach them.
+      # TODO: should we not do this immediately, and give the user more
+      #       control over when such "module defs" are attached?
+      elif name in self.__dataclass_fields__.keys():
+        for suffix, submodule in get_suffix_module_pairs(val):
+          if submodule.parent is _unspecified_parent:
+            submodule.parent = self
+            if submodule.name is not None:
+              raise ValueError(
```

Comment:
As mentioned above I'd prefer a single unified namespace for submodules and parameters, since it minimizes the differences between "compact" and setup() defined data and makes it easier to explain the api (I think).  Happy to talk about this point though.